girder/utility/acl_mixin.py

```diff
@@ -193,9 +201,162 @@ def prefixSearch(self, query, user=None, filters=None, limit=0, offset=0,
+            CommandCursor, it has been augmented with a count function.
+        """
+        # If no user is specified and greater than READ access is requested,
+        # we won't return anything.  So as to always return a cursor for
+        # consistency, we make a query that will return no results
+        if user is None and level is not None and level > AccessType.READ:
+            query = {'__matchnothing': 'nothing'}
+        elif level is not None and (not user or not user['admin']):
+            # If the resourceColl isn't an access controlled model that we
+            # know how to reach, fall back to performing the ordinary query and
+            # then filtering it by permission.  For instance, if a model uses
+            # the acl mixin to get acl from a model that itself uses the acl
+            # mixin, this will return the correct results, but without the
+            # utility of being able perform count().
+            #  Note, this also handles models which use attachedToType and
+            # attachedToId, since self.model(None) will not be an access
+            # controlled model.
+            #  This is also the fall-back for Mongo < 3.4, as those versions do
+            # not support the aggregation steps that are used.
+            if (not isinstance(self.model(self.resourceColl), AccessControlledModel) or
+                    not getattr(self, '_dbserver_version', None) or
+                    getattr(self, '_dbserver_version', None) < (3, 4)):
+                return self._findWithPermissionsFallback(
+                    query, offset, limit, timeout, fields, sort, user, level,
+                    **kwargs)
+
+            query = query or {}
+            initialPipeline = [
+                {'$match': query},
+                {'$lookup': {
+                    'from': self.resourceColl,
+                    'localField': self.resourceParent,
+                    'foreignField': '_id',
+                    'as': '_access'
+                }},
+                {'$match': self.permissionClauses(user, level, '_access.')},
+            ]
+            countPipeline = initialPipeline + [
+                {'$count': 'count'},
+            ]
+            fullPipeline = initialPipeline + [
+                {'$project': {'_access': False}},
+            ]
+            if sort is not None or aggregateSort is not None:
+                fullPipeline.append({'$sort': collections.OrderedDict(sort or aggregateSort)})
+            # limit should immediately follow sort for efficiency
+            if limit:
+                fullPipeline.append({'$limit': limit + (offset or 0)})
+            if offset:
+                fullPipeline.append({'$skip': offset})
+            if fields is not None:
+                if any(isinstance(v, bool) for v in fields.values()):
+                    fullPipeline.append({'$project': fields})
+                else:
+                    fullPipeline.append({'$addFields': fields})
```

Comment:
This `$addFields` case should be cut, as injecting constant data is not something we need to support. We should expect that `fields` is always either, at most:
* A Python `set` of fields that should all be kept
* A Python `dict`, where all values are `bool`
* A Python `dict`, where all values are `int` of values `1` or `0`

I'd be open to further restricting the allowed types of `fields`.

Reply:
This actually gets used as part of the `_textSearchFilters` where the field `_textScore` is set to `{'$meta': 'textScore'}`.  This means that this should probably be changed to:
```
            if fields is not None:
                if isinstance(fields, (abc.Sequence, abc.Set)):
                    fields = dict.fromkeys(fields, 1)
                if any(not isinstance(v, abc.Mapping) for v in fields.values()):
                    fullPipeline.append({'$project': {
                        k: v for k, v in six.iteritems(fields)
                        if not isinstance(v, abc.Mapping)}})
                if any(isinstance(v, abc.Mapping) for v in fields.values()):
                    fullPipeline.append({'$addFields': {
                        k: v for k, v in six.iteritems(fields)
                        if isinstance(v, abc.Mapping)}})
```